text/0066-archival-nodes.md

```diff
@@ -0,0 +1,85 @@
+
+There is a middle ground here, though - instead of having all nodes keep the whole history, or having all nodes discard older blocks, we could have most of the nodes only keeping what is relevant for day-to-day operations, and only a small number of nodes keeping the complete history of the network. These history keepers would be called the "archival nodes".
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Until Fast Sync is introduced, all nodes are effectively archival nodes, as every time a node joins the network, it has to sync the whole blockchain since genesis. Thus, once it becomes able to participate in the network, it already has all the history stored locally.
+
+Fast Sync is going to change that picture. Once it starts being the method used for syncing, the joining nodes will only download some of the switch block headers and the global state from a recent block, ignoring a huge part of the network's history. This means that if a fast syncing node didn't have all the history already, it won't have at least some of it afterwards, either.
+
+Since Fast Sync has to be able to download full data of at least a single block, we propose for the archival nodes to use Fast Sync capabilities to sync each block in the chain separately.
+
+To improve performance, we also propose for the archival sync process to run in the background while the node already participates in the network. Thus, the process would look the following way:
+
+1. Start like a regular joining node.
+2. Fast sync to the latest block like a regular, non-archival node and transition to the "participating" state.
+3. In the background, walk back along the chain, downloading all the blocks down to genesis.
+
+After completing this process, the node will have the complete chain and all historical global state tries.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Going into more details of the process explained above, it would look like this:
+
+1. Launch the node, fast sync to the latest block, transition to the "participating state".
+2. In the background:
+    - Set the current block to the latest one.
+    - Loop:
+        - download the block's global state trie, deploys and finality signatures, if not already present in storage
+        - if the current block height is 0, break
+        - set the current block to the current block's parent
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+With this design, there is a larger risk of losing the history of the network, as the number of complete copies would be dramatically decreased. However, the risk will still be small, and the lowered costs of running a regular nodes seem to be worth it.
```

Comment:
What solutions will be put in place to mitigate this risk ?

Reply:
I don't think this CEP in particular affects this risk - without the change, archival nodes would simply become operational later (after having downloaded all blocks, not directly after they have acquired enough state) instead. The core difference is that with fast sync, it is now _possible_ to run nodes with less history, so hopefully this will be a net increase, enabling those with smaller budgets to run nodes as well.

Arguably this change improves the situation for archival nodes as well, as it limits the cost of running an archival node by making it available sooner --- operators that might have been discouraged by the long startup time for an archival node but otherwise have no issue providing the necessary disk space now no longer need to make that tradeoff.

Reply:
I mean there won't be any advantage to run full archival nodes when this change will be up. So unless good Samaritan keeps some full archival nodes we will effectively lose some history ?

Reply:
There's a number of products that do benefit from having the full history - data aggregators, block explorers, etc.. From the PoV of the protocol/validators, the old history is unnecessary.

Reply:
So does the association / partners are already planning to provide/keep full archival nodes ?

Reply:
Yes.